File: src/main/java/io/femo/http/drivers/server/Http20SocketHandler.java
package io.femo.http.drivers.server;

import io.femo.http.transport.http2.HttpConnection;
import io.femo.http.transport.http2.HttpSettings;
import io.femo.http.transport.http2.frames.SettingsFrame;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.net.Socket;

/**
 * Created by felix on 9/15/16.
 */
public class Http20SocketHandler implements SocketHandler {

    private Logger log = LoggerFactory.getLogger("HTTP");

    private HttpHandlerStack httpHandlerStack;

    public Http20SocketHandler(HttpHandlerStack httpHandlerStack) {
        this.httpHandlerStack = httpHandlerStack;
    }

    @Override
    public void handle(Socket socket) {
        byte[] preface = new byte[24];
        try {
            InputStream inputStream = socket.getInputStream();
            inputStream.read(preface, 0, 24);
            if(!new String(preface).equals("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n")) {
                log.warn("Invalid HTTP/2.0 preface");
                socket.close();
                return;
            }
            HttpConnection httpConnection = new HttpConnection(socket, new HttpSettings(HttpSettings.EndpointType.SERVER));
            httpConnection.handler(httpHandlerStack);
            SettingsFrame settingsFrame = new SettingsFrame(httpConnection.getRemoteSettings());
            httpConnection.enqueueFrame(settingsFrame, null);
            httpConnection.handle();
        } catch (IOException e) {
            log.warn("Error in HTTP/2.0 connection", e);
        }

    }
}

File: src/main/java/org/web25/http/drivers/server/Http20SocketHandler.kt
package org.web25.http.drivers.server

import org.slf4j.LoggerFactory
import org.web25.http.transport.http2.HttpConnection
import org.web25.http.transport.http2.HttpSettings
import org.web25.http.transport.http2.frames.SettingsFrame
import java.io.IOException
import java.net.Socket

/**
 * Created by felix on 9/15/16.
 */
class Http20SocketHandler(private val httpHandlerStack: HttpHandlerStack, private val executorListener: ExecutorListener) : SocketHandler {
    private val log = LoggerFactory.getLogger("HTTP")

    override fun handle(socket: Socket) {
        val preface = ByteArray(24)
        try {
            val inputStream = socket.inputStream
            inputStream.read(preface, 0, 24)
            if (String(preface) != "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n") {
                log.warn("Invalid HTTP/2.0 preface")
                socket.close()
                return
            }
            val httpConnection = HttpConnection(socket, HttpSettings(HttpSettings.EndpointType.SERVER), executorListener)
            httpConnection.handler(httpHandlerStack)
            val settingsFrame = SettingsFrame(httpConnection.remoteSettings)
            httpConnection.enqueueFrame(settingsFrame, null)
            httpConnection.handle()
        } catch (e: IOException) {
            log.warn("Error in HTTP/2.0 connection", e)
        }

    }
}
